File: geekbang/algorithm-models/4.6-二叉搜索树后继模板/inorderSuccessor.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {TreeNode} p
 * @return {TreeNode}
 */
var inorderSuccessor = function(root, p) {
    var findSucc = function(root, val) {
        let ans = null;
        while (root != null) {
            if (val == root.val) {
                if (root.right != null) {
                    let p = root.right;
                    while (p.left != null) p = p.left;
                    return p;
                }
                break;
            }
            if (root.val > val && (ans == null || ans.val > root.val))
                ans = root;
            if (val < root.val) root = root.left; else root = root.right;
        }
        return ans;
    }


    return findSucc(root, p.val);
};

File: geekbang/algorithm-models/4.6-二叉搜索树后继模板/inorderSuccessor.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def inorderSuccessor(self, root: TreeNode, p: TreeNode) -> TreeNode:
        return self.findSucc(root, p.val)


    def findSucc(self, root, val):
        ans = None
        while root:
            if val == root.val:
                if root.right:
                    p = root.right
                    while p.left:
                        p = p.left
                    return p
                break
            if root.val > val:
                if ans is None or ans.val > root.val:
                    ans = root
            if val < root.val:
                root = root.left
            else:
                root = root.right
        return ans
